pkg/controller/clone-controller.go

```diff
@@ -0,0 +1,269 @@
+	for c.ProcessNextPodItem() {
+	}
+}
+
+func (c *CloneController) runPVCWorkers() {
+	for c.ProcessNextPvcItem() {
+	}
+}
+
+// ProcessNextPodItem gets the next pod key from the queue and verifies that it was created by the
+// controller. If not the key is discarded; otherwise, the pod object is passed to processPodItem.
+// Note: pods are already filtered by label "host-assisted-cloning".
+func (c *CloneController) ProcessNextPodItem() bool {
+	key, shutdown := c.podQueue.Get()
+	if shutdown {
+		return false
+	}
+	defer c.podQueue.Done(key)
+	pod, err := c.podFromKey(key)
+	if err != nil {
+		c.forgetKey(key, fmt.Sprintf("ProcessNextPodItem: unable to get pod from key %v: %v", key, err))
+		return true
+	}
+	if !metav1.HasAnnotation(pod.ObjectMeta, AnnCloningCreatedBy) {
+		c.forgetKey(key, fmt.Sprintf("ProcessNextPodItem: pod %q does not have annotation %q", key, AnnCloningCreatedBy))
+		return true
+	}
+	glog.V(Vdebug).Infof("ProcessNextPodItem: next pod to process: %s\n", key)
+	err = c.processPodItem(pod)
+	if err != nil { // processPodItem errors may not have been logged so log here
+		glog.Errorf("error processing pod %q: %v", key, err)
+		return true
+	}
+	return c.forgetKey(key, fmt.Sprintf("ProcessNextPodItem: processing pod %q completed", key))
+}
+
+// processPodItem verifies that the passed in pod is genuine and, if so, annotates the Phase
+// of the pod in the PVC to indicate the status of the cloning process.
+func (c *CloneController) processPodItem(pod *v1.Pod) error {
+	// verify that this pod has the expected pvc name
+	var pvcKey string
+	for _, vol := range pod.Spec.Volumes {
+		if vol.Name == ImagePathName {
+			pvcKey = fmt.Sprintf("%s/%s", pod.Namespace, vol.PersistentVolumeClaim.ClaimName)
+			glog.V(Vadmin).Infof("pod \"%s/%s\" has volume matching claim %q\n", pod.Namespace, pod.Name, pvcKey)
+			break
+		}
+	}
+	if len(pvcKey) == 0 {
+		// If this block is ever reached, something has gone very wrong.  The pod should ALWAYS be created with the volume.
+		// A missing volume would most likely indicate a pod that has been created manually, but also incorrectly defined.
+		return errors.Errorf("Pod does not contain volume %q", ImagePathName)
+	}
+
+	glog.V(Vdebug).Infof("processPodItem: getting pvc for key %q", pvcKey)
+	pvc, err := c.pvcFromKey(pvcKey)
+	if err != nil {
+		return errors.WithMessage(err, "could not retrieve pvc from cache")
+	}
+	// see if pvc's pod phase anno needs to be added/updated
+	phase := string(pod.Status.Phase)
+	if !checkIfAnnoExists(pvc, AnnPodPhase, phase) {
+		pvc, err = setPVCAnnotation(c.clientset, pvc, AnnPodPhase, phase)
+		if err != nil {
+			return errors.WithMessage(err, fmt.Sprintf("could not set annotation \"%s: %s\" on pvc %q", AnnPodPhase, phase, pvc.Name))
+			glog.V(Vdebug).Infof("processPodItem: pod phase %q annotated in pvc %q", pod.Status.Phase, pvcKey)
+		}
+		if phase == "Succeeded" {
+			pvc, err = setPVCAnnotation(c.clientset, pvc, AnnCloneOf, "true")
```

Comment:
couldn't this code be updating the golden/source pvc? We don't want that pvc to be updated with "cloneof". I guess it could be updated to say it's the source of a clone but not sure we need this.

Reply:
The clone controller should not alter the golden PVC's annotations at all.  Currently, this will overwrite the import pod's status annotation in the PVC, which kubevirt relies on for determining the completion of an import.

Reply:
you right. working to fix that.